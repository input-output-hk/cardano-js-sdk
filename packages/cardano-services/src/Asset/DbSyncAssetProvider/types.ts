export interface LastMintTxModel {
  tx_hash: Buffer;
}

export interface MultiAssetModel {
  fingerprint: string;
  id: string;
}

export interface MultiAssetHistoryModel {
  hash: Buffer;
  quantity: string;
}

export interface MultiAssetQuantitiesModel {
  count: string;
  sum: string;
}
