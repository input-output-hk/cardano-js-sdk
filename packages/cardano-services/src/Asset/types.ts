import { Asset, Cardano } from '@cardano-sdk/core';
import { Shutdown } from '@cardano-sdk/util';

/**
 * Cardano.AssetId as an object with `policyId` and `name`
 */
export type AssetPolicyIdAndName = Pick<Asset.AssetInfo, 'name' | 'policyId'>;

/**
 * Service to get CIP-25 NFT metadata for a given asset
 */
export interface NftMetadataService {
  /**
   * Get CIP-25 NFT metadata for a given asset
   *
   * @returns CIP-25 NFT metadata for NFTs, `null` for assets that are not NFTs
   */
  getNftMetadata(asset: AssetPolicyIdAndName): Promise<Asset.NftMetadata | null>;
}

/**
 * Service to get CIP-38? token metadata for a given subject
 */
export interface TokenMetadataService extends Shutdown {
  /**
   * Get CIP-38? token metadata for a given subject
   *
   * @returns CIP-38? token metadata, `null` if not found
   */
  getTokenMetadata(assetIds: Cardano.AssetId[]): Promise<(Asset.TokenMetadata | null)[]>;
}
