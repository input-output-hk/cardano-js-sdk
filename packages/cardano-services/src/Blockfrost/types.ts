export interface CurrentEpochModel {
  epoch_no: number;
}

export interface FirstUpdateAfterBlock {
  epoch_no: number;
}

export interface FirstUpdateAfterBlockModel {
  active_epoch_no: string;
}

export interface LastRetireModel {
  block_no: number;
  retiring_epoch: number;
}

export interface PoolsModel {
  id: string;
  view: string;
}
