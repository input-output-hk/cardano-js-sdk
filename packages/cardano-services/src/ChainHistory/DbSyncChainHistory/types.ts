import { Cardano } from '@cardano-sdk/core';

export type TransactionDataMap<T> = Map<Cardano.TransactionId, T>;
export type TxOutTokenMap = Map<string, Cardano.TokenMap>;
export type TxTokenMap = TransactionDataMap<Cardano.TokenMap>;

export interface BlockModel {
  block_no: number;
  epoch_no: number;
  epoch_slot_no: number;
  hash: Buffer;
  next_block: Buffer | null;
  previous_block: Buffer | null;
  size: number;
  slot_leader_hash: Buffer;
  slot_leader_pool: string | null;
  slot_no: string;
  time: string;
  tx_count: string;
  vrf: string;
}

export interface BlockOutputModel {
  fees: string;
  output: string;
  hash: Buffer;
}

export interface TipModel {
  block_no: number;
  hash: Buffer;
  slot_no: number;
}

export interface TxModel {
  id: Buffer;
  index: number;
  size: number;
  fee: string;
  invalid_before: string | null;
  invalid_hereafter: string | null;
  block_no: number;
  block_hash: Buffer;
  block_slot_no: number;
}

export interface TxInputModel {
  address: string;
  id: string;
  index: number;
  tx_input_id: Buffer;
  tx_source_id: Buffer;
}

export interface TxInput {
  address: Cardano.Address;
  id: string;
  index: number;
  txInputId: Cardano.TransactionId;
  txSourceId: Cardano.TransactionId;
}

export interface TxOutputModel {
  address: string;
  coin_value: string;
  datum?: Buffer | null;
  id: string;
  index: number;
  tx_id: Buffer;
}

export interface TxOutput extends Cardano.TxOut {
  txId: Cardano.TransactionId;
  index: number;
}

export interface MultiAssetModel {
  asset_name: Buffer;
  fingerprint: string;
  policy_id: Buffer;
  quantity: string;
  tx_id: Buffer;
}

export interface TxOutMultiAssetModel extends MultiAssetModel {
  tx_out_id: string;
}

export interface WithdrawalModel {
  quantity: string;
  tx_id: Buffer;
  stake_address: string;
}

export interface RedeemerModel {
  index: number;
  purpose: string;
  script_hash: Buffer;
  unit_mem: string;
  unit_steps: string;
  tx_id: Buffer;
}

export interface CertificateModel {
  cert_index: number;
  tx_id: Buffer;
}
export type WithCertIndex<T extends Cardano.Certificate> = T & { cert_index: number };
export type WithCertType<T extends CertificateModel> = T & {
  type: 'retire' | 'register' | 'mir' | 'stake' | 'delegation';
};

export interface PoolRetireCertModel extends CertificateModel {
  retiring_epoch: number;
  pool_id: string;
}
export interface PoolRegisterCertModel extends CertificateModel {
  pool_id: string;
}

export interface MirCertModel extends CertificateModel {
  amount: string;
  pot: 'reserve' | 'treasury';
  address: string;
}
export interface StakeCertModel extends CertificateModel {
  address: string;
  registration: boolean;
}

export interface DelegationCertModel extends CertificateModel {
  address: string;
  pool_id: string;
}
