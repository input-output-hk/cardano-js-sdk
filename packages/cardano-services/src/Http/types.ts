import { Options } from 'body-parser';
import expressPromBundle from 'express-prom-bundle';
import net from 'net';

export type ServiceHealth = {
  ok: boolean;
  name: string;
};

export type ServicesHealthCheckResponse = {
  ok: boolean;
  services: ServiceHealth[];
};

export type ServerMetadata = {
  lastModified?: number;
  lastModifiedDate?: string;
  rev?: string;
  shortRev?: string;
  extra?: JSON;
  startupTime: number;
};

export type HttpServerConfig = {
  metrics?: {
    enabled: boolean;
    options?: expressPromBundle.Opts;
  };
  bodyParser?: {
    limit?: Options['limit'];
  };
  name?: string;
  meta?: ServerMetadata;
  listen: net.ListenOptions;
  allowedOrigins?: string[];
};

export type BuildInfo = Omit<ServerMetadata, 'extra' | 'startupTime'>;
