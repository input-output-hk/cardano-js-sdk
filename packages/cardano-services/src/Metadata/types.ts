import { Cardano } from '@cardano-sdk/core';

export interface TxMetadataService {
  queryTxMetadataByHashes(hashes: Cardano.TransactionId[]): Promise<Map<Cardano.TransactionId, Cardano.TxMetadata>>;
}

export interface TxMetadataModel {
  key: string;
  json_value: { [k: string]: unknown };
  tx_id: Buffer;
}
