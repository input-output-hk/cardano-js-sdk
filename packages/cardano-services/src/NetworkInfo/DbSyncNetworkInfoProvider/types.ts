import { Cardano } from '@cardano-sdk/core';

export interface CirculatingSupplyModel {
  circulating_supply: string;
}

export interface TotalSupplyModel {
  total_supply: string;
}

export interface ActiveStakeModel {
  active_stake: string;
}

export interface LiveStakeModel {
  live_stake: string;
}

export interface EpochModel {
  no: number;
}

export interface CostModelsParamModel {
  PlutusV1?: Cardano.CostModel;
  PlutusV2?: Cardano.CostModel;
  PlutusV3?: Cardano.CostModel;
}

export interface ProtocolParamsModel {
  coins_per_utxo_size: string;
  max_tx_size: number;
  max_val_size: string;
  min_pool_cost: string;
  key_deposit: string;
  pool_deposit: string;
  protocol_major: number;
  protocol_minor: number;
  min_fee_a: number;
  min_fee_b: number;
  max_collateral_inputs: number;
  max_block_size: number;
  max_bh_size: number;
  optimal_pool_count: number;
  influence: number;
  monetary_expand_rate: number;
  treasury_growth_rate: number;
  decentralisation: number;
  collateral_percent: number;
  price_mem: number;
  price_step: number;
  max_tx_ex_mem: string;
  max_tx_ex_steps: string;
  max_block_ex_mem: string;
  max_block_ex_steps: string;
  max_epoch: number;
  costs: CostModelsParamModel | null;
  pool_voting_thresholds: Cardano.PoolVotingThresholds;
  drep_voting_thresholds: Cardano.DelegateRepresentativeThresholds;
  min_committee_size: number;
  committee_term_limit: number;
  governance_action_validity_period: number;
  governance_action_deposit: number;
  drep_deposit: number;
  drep_inactivity_period: number;
}
