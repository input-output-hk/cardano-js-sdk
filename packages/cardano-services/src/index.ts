export * from './Asset';
export * from './Http';
export * from './InMemoryCache';
export * from './Program';
export * from './RunnableModule';
export * from './StakePool';
export * from './TxSubmit';
export * from './Utxo';
export * from './ChainHistory';
export * from './Rewards';
export * as errors from './errors';
export * as util from './util';
