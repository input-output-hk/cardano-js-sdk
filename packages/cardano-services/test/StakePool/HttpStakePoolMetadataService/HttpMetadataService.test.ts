/* eslint-disable max-len */
import { Cardano, ProviderError, ProviderFailure } from '@cardano-sdk/core';
import { ExtMetadataFormat } from '../../../src/StakePool/types';
import { adaPoolsExtMetadataMock, cip6ExtMetadataMock, mainExtMetadataMock } from './mocks';
import { createGenericMockServer, logger } from '@cardano-sdk/util-dev';
import { createHttpStakePoolExtMetadataService } from '../../../src';
import url from 'url';

export const mockPoolExtMetadataServer = createGenericMockServer((handler) => async (req, res) => {
  const result = handler(req);

  res.setHeader('Content-Type', 'application/json');

  if (result.body) {
    res.statusCode = result.code || 200;
    return res.end(JSON.stringify(result.body));
  }

  const reqUrl = url.parse(req.url!).pathname;

  if (reqUrl === `/${ExtMetadataFormat.AdaPools}`) {
    return res.end(JSON.stringify(adaPoolsExtMetadataMock));
  } else if (reqUrl === `/${ExtMetadataFormat.CIP6}`) {
    return res.end(JSON.stringify(cip6ExtMetadataMock));
  }

  return res.end(JSON.stringify(adaPoolsExtMetadataMock));
});

describe('StakePoolExtMetadataService', () => {
  describe('healthy state', () => {
    let closeMock: () => Promise<void> = jest.fn();
    let serverUrl = '';
    const extendedMetadataService = createHttpStakePoolExtMetadataService(logger);

    beforeAll(async () => {
      ({ closeMock, serverUrl } = await mockPoolExtMetadataServer(() => ({})));
    });

    afterAll(async () => {
      await closeMock();
    });

    it('returns ada pools format when extended key is present in the metadata', async () => {
      const extMetadata: Cardano.StakePoolMetadata = {
        ...mainExtMetadataMock(),
        extended: `${serverUrl}/${ExtMetadataFormat.AdaPools}`
      };
      const result = await extendedMetadataService.getStakePoolExtendedMetadata(extMetadata);

      expect(result).not.toBeNull();
      expect(result).toMatchSnapshot();
    });

    it('returns CIP-6 format when extDataUrl is present in the metadata', async () => {
      const extMetadata: Cardano.StakePoolMetadata = {
        ...mainExtMetadataMock(),
        extDataUrl: `${serverUrl}/${ExtMetadataFormat.CIP6}`
      };
      const result = await extendedMetadataService.getStakePoolExtendedMetadata(extMetadata);

      expect(result).not.toBeNull();
      expect(result).toMatchSnapshot();
    });

    it('returns CIP-6 format with priority when the metadata including both extended properties', async () => {
      const extMetadata: Cardano.StakePoolMetadata = {
        ...mainExtMetadataMock(),
        extDataUrl: `${serverUrl}/${ExtMetadataFormat.CIP6}`,
        extended: `${serverUrl}/${ExtMetadataFormat.AdaPools}`
      };
      const result = await extendedMetadataService.getStakePoolExtendedMetadata(extMetadata);

      expect(result).not.toBeNull();
      expect(result?.serial).toBeDefined();
    });
  });

  describe('error cases are correctly handled', () => {
    const extendedMetadataService = createHttpStakePoolExtMetadataService(logger);
    let closeMock: () => Promise<void> = jest.fn();
    let serverUrl: string;

    beforeEach(() => (closeMock = jest.fn()));

    afterEach(async () => await closeMock());

    it('invalid CIP-6 response format', async () => {
      const invalidCip6ResponseFormat = { pool1: { ...cip6ExtMetadataMock.pool }, serial: 12_345 };
      ({ closeMock, serverUrl } = await mockPoolExtMetadataServer(() => ({ body: invalidCip6ResponseFormat })));

      const extMetadata: Cardano.StakePoolMetadata = {
        ...mainExtMetadataMock(),
        extDataUrl: `${serverUrl}/${ExtMetadataFormat.CIP6}`
      };

      await expect(extendedMetadataService.getStakePoolExtendedMetadata(extMetadata)).rejects.toThrow(
        new ProviderError(
          ProviderFailure.InvalidResponse,
          'instance requires property "pool"',
          'Extended metadata JSON format validation failed against the corresponding schema for correctness'
        )
      );
    });

    it('invalid AP response format', async () => {
      const invalidAdaPoolsResponseFormat = { info1: { ...adaPoolsExtMetadataMock.info } };
      ({ closeMock, serverUrl } = await mockPoolExtMetadataServer(() => ({ body: invalidAdaPoolsResponseFormat })));

      const extMetadata: Cardano.StakePoolMetadata = {
        ...mainExtMetadataMock(),
        extended: `${serverUrl}/${ExtMetadataFormat.AdaPools}`
      };

      await expect(extendedMetadataService.getStakePoolExtendedMetadata(extMetadata)).rejects.toThrow(
        new ProviderError(
          ProviderFailure.InvalidResponse,
          'instance requires property "info"',
          'Extended metadata JSON format validation failed against the corresponding schema for correctness'
        )
      );
    });

    it('internal server error', async () => {
      let alreadyCalled = false;
      const handler = () => {
        if (alreadyCalled) return { body: {}, code: 500 };
        alreadyCalled = true;

        return {
          body: cip6ExtMetadataMock
        };
      };

      ({ closeMock, serverUrl } = await mockPoolExtMetadataServer(handler));

      const extMetadata: Cardano.StakePoolMetadata = {
        ...mainExtMetadataMock(),
        extDataUrl: `${serverUrl}/${ExtMetadataFormat.CIP6}`
      };

      const result = await extendedMetadataService.getStakePoolExtendedMetadata(extMetadata);
      expect(result).toBeDefined();

      await expect(extendedMetadataService.getStakePoolExtendedMetadata(extMetadata)).rejects.toThrow(
        new ProviderError(
          ProviderFailure.ConnectionFailure,
          null,
          `StakePoolExtMetadataService failed to fetch extended metadata from ${serverUrl}/${ExtMetadataFormat.CIP6} due to connection error`
        )
      );
    });

    it('resource not found server error', async () => {
      let alreadyCalled = false;
      const handler = () => {
        if (alreadyCalled) return { body: {}, code: 404 };
        alreadyCalled = true;

        return {
          body: cip6ExtMetadataMock
        };
      };

      ({ closeMock, serverUrl } = await mockPoolExtMetadataServer(handler));

      const extMetadata: Cardano.StakePoolMetadata = {
        ...mainExtMetadataMock(),
        extDataUrl: `${serverUrl}/${ExtMetadataFormat.CIP6}`
      };

      const result = await extendedMetadataService.getStakePoolExtendedMetadata(extMetadata);
      expect(result).toBeDefined();

      await expect(extendedMetadataService.getStakePoolExtendedMetadata(extMetadata)).rejects.toThrow(
        new ProviderError(
          ProviderFailure.NotFound,
          null,
          `StakePoolExtMetadataService failed to fetch extended metadata from ${serverUrl}/${ExtMetadataFormat.CIP6} due to resource not found`
        )
      );
    });
  });
});
