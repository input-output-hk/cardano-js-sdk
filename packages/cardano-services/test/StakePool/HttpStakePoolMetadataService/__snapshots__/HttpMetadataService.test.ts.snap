// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StakePoolExtMetadataService healthy state returns CIP-6 format when extDataUrl is present in the metadata 1`] = `
Object {
  "pool": Object {
    "contact": Object {
      "email": "help@pooldomain.org",
      "facebook": "demopool",
      "feed": "https://demopool.com/xml/poolrss.xml",
      "github": "demopool",
      "primary": "email",
      "telegram": "demopool",
      "twitter": "demopool",
    },
    "id": "4a2e3b7f4a78ff1452b91329a7673c77c98ca96dece7b55c37869502",
    "itn": Object {
      "owner": "ed25519_pk1...",
      "witness": "ed25519_sig1...",
    },
    "location": "DE",
    "media_assets": Object {
      "color_bg": "#RRGGBB",
      "color_fg": "#RRGGBB",
      "icon_png_64x64": "https://mydemopool.com/icon.png",
      "logo_png": "https://mydemopool.com/logo.png",
      "logo_svg": "https://mydemopool.com/logo.svg",
    },
    "status": "active",
  },
  "serial": 2020072001,
}
`;

exports[`StakePoolExtMetadataService healthy state returns ada pools format when extended key is present in the metadata 1`] = `
Object {
  "pool": Object {
    "about": Object {
      "company": "IOHK",
      "me": "Cardano",
      "server": "Raspberry Pi Hosted at my Home",
    },
    "company": Object {
      "addr": "Street, Number",
      "city": "London",
      "company_id": "123456789",
      "country": "England",
      "name": "Company Name",
      "vat_id": "GB123456789",
    },
    "contact": Object {
      "discord": "",
      "facebook": "",
      "github": "",
      "telegram": "adapools_ops",
      "twitch": "",
      "twitter": "adapools_org",
      "youtube": "",
    },
    "location": "London, England",
    "media_assets": Object {
      "icon_png_64x64": "http(s) url to pool icon; png format; not animated; max 40kb",
      "logo_png": "http(s) url to pool logo; png format; not animated; max 50kb",
    },
  },
}
`;
